src/encoder/encoderopus.cpp

```diff
@@ -0,0 +1,439 @@
+    // - Comment list: dynamic size
+    QMapIterator<QString, QString> iter(m_opusComments);
+    while(iter.hasNext()) {
+        iter.next();
+        QString comment = iter.key() + "=" + iter.value();
+        QByteArray commentBytes = comment.toUtf8();
+        int commentBytesLength = commentBytes.size();
+
+        // One comment is:
+        // - 4 bytes of string length
+        // - string data
+
+        // Add comment length field and data to comments "list"
+        for (int x = 0; x < 4; x++) {
+            unsigned char fieldValue = (commentBytesLength >> (x*8)) & 0xFF;
+            combinedComments.append(fieldValue);
+        }
+        combinedComments.append(commentBytes);
+
+        // Don't forget to include this comment in the overall size calculation
+        frameSize += (4 + commentBytesLength);
+        commentCount++;
+    }
+
+    // == Actual frame building ==
+    unsigned char* data = new unsigned char[frameSize]();
+    int pos = 0; // Current position
+
+    // Magic signature (8 bytes)
+    memcpy(data + pos, "OpusTags", 8);
+    pos += 8;
+
+    // Vendor string (mandatory)
+    // length field (4 bytes, little-endian) + actual string
+    // Write length field
+    for (int x = 0; x < 4; x++) {
+        data[pos++] = (vendorStringLength >> (x*8)) & 0xFF;
+    }
+    // Write string
+    memcpy(data + pos, vendorString, vendorStringLength);
+    pos += vendorStringLength;
+
+    // Number of comments (4 bytes, little-endian)
+    for (int x = 0; x < 4; x++) {
+        data[pos++] = (commentCount >> (x*8)) & 0xFF;
+    }
+
+    // Comment list (dynamic size)
+    int commentListLength = combinedComments.size();
+    memcpy(data + pos, combinedComments.constData(), commentListLength);
+    pos += commentListLength;
+
+    // Push finished tags frame to stream
+    ogg_packet packet;
+    packet.b_o_s = 0;
+    packet.e_o_s = 0;
+    packet.granulepos = 0;
+    packet.packetno = m_packetNumber++;
+    packet.packet = data;
+    packet.bytes = frameSize;
+
+    kLogger.debug() << data;
+
+    ogg_stream_packetin(&m_oggStream, &packet);
+    delete[] data;
+}
+
+void EncoderOpus::encodeBuffer(const CSAMPLE *samples, const int size) {
+    if (!m_pOpus) {
+        return;
+    }
+
+    int writeRequired = size;
+    int writeAvailable = m_pFifoBuffer->writeAvailable();
+    if (writeRequired > writeAvailable) {
+        kLogger.warning() << "FIFO buffer too small, loosing samples!"
+                          << "required:" << writeRequired
+                          << "; available: " << writeAvailable;
+    }
+
+    int writeCount = math_min(writeRequired, writeAvailable);
+    if (writeCount > 0) {
+        m_pFifoBuffer->write(samples, writeCount);
+    }
+
+    processFIFO();
+}
+
+void EncoderOpus::processFIFO() {
+    while (m_pFifoBuffer->readAvailable() >= m_readRequired) {
+        m_pFifoBuffer->read(m_pFifoChunkBuffer, m_readRequired);
+
+        int samplesPerChannel = m_readRequired / m_channels;
```

Comment:
Here you might drop some samples. I would at least check with a debug assertion that indeed (m_readRequired % m_channels) == 0 as expect. This will safe you a lot of debugging time if your assumption is wrong, maybe because of some programming error somewhere else.

Reply:
Should be fixed: https://github.com/Palakis/mixxx/blob/opus-encoder/src/encoder/encoderopus.cpp#L344-L346